packages/x-date-pickers/src/internals/hooks/useField/buildSectionsFromFormat.ts

```diff
@@ -202,56 +202,53 @@ const buildSections = <TDate extends PickerValidDate>(
   const sections: FieldSection[] = [];
   let startSeparator: string = '';
 
-  // This RegExp test if the beginning of a string corresponds to a supported token
-  const isTokenStartRegExp = new RegExp(
-    `^(${Object.keys(utils.formatTokenMap)
-      .sort((a, b) => b.length - a.length) // Sort to put longest word first
-      .join('|')})`,
-    'g', // used to get access to lastIndex state
-  );
+  // This RegExp tests if the beginning of a string corresponds to a supported token
+  const isTokenStartRegExpStr = `^(${Object.keys(utils.formatTokenMap)
```

Comment:
Not sure why but my code does not work if I re-use the same `RegExp` object
So I just created the string outside of the loop and the `RegExp` object inside it.